gstreamer/src/buffer_serde.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (C) 2018 François Laignel <fengalin@free.fr>
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use serde::de::{Deserialize, Deserializer};
+use serde::ser;
+use serde::ser::{Serialize, SerializeStruct, Serializer};
+use serde_bytes::{ByteBuf, Bytes};
+
+use Buffer;
+use BufferFlags;
+use BufferRef;
+use ClockTime;
+
+impl<'a> Serialize for BufferRef {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        let mut buffer = serializer.serialize_struct("Buffer", 6)?;
+        buffer.serialize_field("pts", &self.get_pts())?;
+        buffer.serialize_field("dts", &self.get_dts())?;
+        buffer.serialize_field("duration", &self.get_duration())?;
+        buffer.serialize_field("offset", &self.get_offset())?;
+        buffer.serialize_field("offset_end", &self.get_offset_end())?;
+        buffer.serialize_field("flags", &self.get_flags())?;
+        {
+            let data = self
+                .map_readable()
+                .ok_or_else(|| ser::Error::custom("Couldn't map `buffer` as readable"))?;
+            buffer.serialize_field("buffer", &Bytes::new(data.as_slice()))?;
+        }
+        buffer.end()
+    }
+}
+
+impl<'a> Serialize for Buffer {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        self.as_ref().serialize(serializer)
+    }
+}
+
+#[derive(Deserialize)]
+struct BufferDe {
+    pts: ClockTime,
+    dts: ClockTime,
+    duration: ClockTime,
+    offset: u64,
+    offset_end: u64,
+    flags: BufferFlags,
+    buffer: ByteBuf,
+}
```

Comment:
One thing I was wondering is if it would make sense to use that struct also for serialization and let both directions be derived and then only map in both directions. That way the struct definition does not have to be kept in sync with the separate serialization steps above.

Reply:
I didn't do that because I wanted to avoid copies, especially for the `buffer` field. I need to check again, but I don't think I can make it borrow the buffer during ser and own it in de. But it's true that consistency would be verified at compile time with your proposal. Other types such as `DateTime` would benefit from this approach without much extra overhead though.

Reply:
Could `Cow` be used for that?

Reply:
Sure! Thanks.

Reply:
Can it really? That's great :)

Reply:
After trying a bit, `Cow` doesn't seem to help. We would need to use either `Cow<'a, Bytes>` or `Cow<'a, ByteBuf>`. In the `Bytes` case, we need to copy the `[u8]` during deserialization because the `Deserializer` owns it and in the `ByteBuf` case, we need to copy before serialization due to its owning by nature. My proposal, although asymmetrical, avoids unneeded copy AFAICT.

Reply:
I guess it could be an `Either<ByteBuf, Bytes>` but at that point it just becomes silly. Let's keep it as is then, thanks for checking :)

Reply:
I'll try it later, however since it wouldn't change the actual behaviour, I propose not to block the PR on this. I'll feel released somewhat when it is merged.

Reply:
Yes let's keep as-is and maybe add a TODO comment there for the future